<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.12.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran_variances: scran_variances Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran_variances
   </div>
   <div id="projectbrief">Model per-gene variance in expression</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.12.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">scran_variances Namespace Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Variance modelling for single-cell expression data.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Options for <code><a class="el" href="#a836ab33e2b96134e9b3ad47df42aeb5e">choose_highly_variable_genes()</a></code>.  <a href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1FitVarianceTrendOptions.html">FitVarianceTrendOptions</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Options for <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code>.  <a href="structscran__variances_1_1FitVarianceTrendOptions.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1FitVarianceTrendResults.html">FitVarianceTrendResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Results of <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code>.  <a href="structscran__variances_1_1FitVarianceTrendResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1FitVarianceTrendWorkspace.html">FitVarianceTrendWorkspace</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Workspace for <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code>.  <a href="structscran__variances_1_1FitVarianceTrendWorkspace.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedBuffers.html">ModelGeneVariancesBlockedBuffers</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Buffers for <code><a class="el" href="#a34ffafc531b113e60ae57b3b5f8256be">model_gene_variances_blocked()</a></code>.  <a href="structscran__variances_1_1ModelGeneVariancesBlockedBuffers.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedResults.html">ModelGeneVariancesBlockedResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Results of <code><a class="el" href="#a34ffafc531b113e60ae57b3b5f8256be">model_gene_variances_blocked()</a></code>.  <a href="structscran__variances_1_1ModelGeneVariancesBlockedResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBuffers.html">ModelGeneVariancesBuffers</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Buffers for <code><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances()</a></code> and friends.  <a href="structscran__variances_1_1ModelGeneVariancesBuffers.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Options for <code><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances()</a></code> and friends.  <a href="structscran__variances_1_1ModelGeneVariancesOptions.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran__variances_1_1ModelGeneVariancesResults.html">ModelGeneVariancesResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Results of <code><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances()</a></code>.  <a href="structscran__variances_1_1ModelGeneVariancesResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:a836ab33e2b96134e9b3ad47df42aeb5e" id="r_a836ab33e2b96134e9b3ad47df42aeb5e"><td class="memTemplParams" colspan="2">template&lt;typename Stat_ , typename Bool_ &gt; </td></tr>
<tr class="memitem:a836ab33e2b96134e9b3ad47df42aeb5e"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#a836ab33e2b96134e9b3ad47df42aeb5e">choose_highly_variable_genes</a> (const std::size_t n, const Stat_ *const statistic, Bool_ *const output, const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;options)</td></tr>
<tr class="separator:a836ab33e2b96134e9b3ad47df42aeb5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad2e99f0413753c20e33c217e273ee120" id="r_ad2e99f0413753c20e33c217e273ee120"><td class="memTemplParams" colspan="2">template&lt;typename Bool_  = char, typename Stat_ &gt; </td></tr>
<tr class="memitem:ad2e99f0413753c20e33c217e273ee120"><td class="memTemplItemLeft" align="right" valign="top">std::vector&lt; Bool_ &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#ad2e99f0413753c20e33c217e273ee120">choose_highly_variable_genes</a> (const std::size_t n, const Stat_ *const statistic, const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;options)</td></tr>
<tr class="separator:ad2e99f0413753c20e33c217e273ee120"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a526febe6eb07121cdd6649a938992ad2" id="r_a526febe6eb07121cdd6649a938992ad2"><td class="memTemplParams" colspan="2">template&lt;typename Index_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a526febe6eb07121cdd6649a938992ad2"><td class="memTemplItemLeft" align="right" valign="top">std::vector&lt; Index_ &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#a526febe6eb07121cdd6649a938992ad2">choose_highly_variable_genes_index</a> (const Index_ n, const Stat_ *const statistic, const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;options)</td></tr>
<tr class="separator:a526febe6eb07121cdd6649a938992ad2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a774f5059db93cedb24f461201f3131fd" id="r_a774f5059db93cedb24f461201f3131fd"><td class="memTemplParams" colspan="2">template&lt;typename Float_ &gt; </td></tr>
<tr class="memitem:a774f5059db93cedb24f461201f3131fd"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend</a> (const std::size_t n, const Float_ *const mean, const Float_ *const variance, Float_ *const fitted, Float_ *const residuals, <a class="el" href="structscran__variances_1_1FitVarianceTrendWorkspace.html">FitVarianceTrendWorkspace</a>&lt; Float_ &gt; &amp;workspace, const <a class="el" href="structscran__variances_1_1FitVarianceTrendOptions.html">FitVarianceTrendOptions</a> &amp;options)</td></tr>
<tr class="separator:a774f5059db93cedb24f461201f3131fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada30bbfa19a0c28c23fb69f61355ea54" id="r_ada30bbfa19a0c28c23fb69f61355ea54"><td class="memTemplParams" colspan="2">template&lt;typename Float_ &gt; </td></tr>
<tr class="memitem:ada30bbfa19a0c28c23fb69f61355ea54"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran__variances_1_1FitVarianceTrendResults.html">FitVarianceTrendResults</a>&lt; Float_ &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#ada30bbfa19a0c28c23fb69f61355ea54">fit_variance_trend</a> (const std::size_t n, const Float_ *const mean, const Float_ *const variance, const <a class="el" href="structscran__variances_1_1FitVarianceTrendOptions.html">FitVarianceTrendOptions</a> &amp;options)</td></tr>
<tr class="separator:ada30bbfa19a0c28c23fb69f61355ea54"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34ffafc531b113e60ae57b3b5f8256be" id="r_a34ffafc531b113e60ae57b3b5f8256be"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a34ffafc531b113e60ae57b3b5f8256be"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#a34ffafc531b113e60ae57b3b5f8256be">model_gene_variances_blocked</a> (const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;mat, const Block_ *const block, const <a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedBuffers.html">ModelGeneVariancesBlockedBuffers</a>&lt; Stat_ &gt; &amp;buffers, const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;options)</td></tr>
<tr class="separator:a34ffafc531b113e60ae57b3b5f8256be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af48891a8979a6e46b37aaf7c9bc027cb" id="r_af48891a8979a6e46b37aaf7c9bc027cb"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:af48891a8979a6e46b37aaf7c9bc027cb"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances</a> (const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;mat, <a class="el" href="structscran__variances_1_1ModelGeneVariancesBuffers.html">ModelGeneVariancesBuffers</a>&lt; Stat_ &gt; buffers, const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;options)</td></tr>
<tr class="separator:af48891a8979a6e46b37aaf7c9bc027cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab7571d9531d24fb47fa0874ff5039725" id="r_ab7571d9531d24fb47fa0874ff5039725"><td class="memTemplParams" colspan="2">template&lt;typename Stat_  = double, typename Value_ , typename Index_ &gt; </td></tr>
<tr class="memitem:ab7571d9531d24fb47fa0874ff5039725"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran__variances_1_1ModelGeneVariancesResults.html">ModelGeneVariancesResults</a>&lt; Stat_ &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#ab7571d9531d24fb47fa0874ff5039725">model_gene_variances</a> (const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;mat, const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;options)</td></tr>
<tr class="separator:ab7571d9531d24fb47fa0874ff5039725"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af13889165f096e5af5db48a6046e2d27" id="r_af13889165f096e5af5db48a6046e2d27"><td class="memTemplParams" colspan="2">template&lt;typename Stat_  = double, typename Value_ , typename Index_ , typename Block_ &gt; </td></tr>
<tr class="memitem:af13889165f096e5af5db48a6046e2d27"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedResults.html">ModelGeneVariancesBlockedResults</a>&lt; Stat_ &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="#af13889165f096e5af5db48a6046e2d27">model_gene_variances_blocked</a> (const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;mat, const Block_ *const block, const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;options)</td></tr>
<tr class="separator:af13889165f096e5af5db48a6046e2d27"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Variance modelling for single-cell expression data. </p>
</div><h2 class="groupheader">Function Documentation</h2>
<a id="a836ab33e2b96134e9b3ad47df42aeb5e" name="a836ab33e2b96134e9b3ad47df42aeb5e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a836ab33e2b96134e9b3ad47df42aeb5e">&#9670;&#160;</a></span>choose_highly_variable_genes() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Stat_ , typename Bool_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void scran_variances::choose_highly_variable_genes </td>
          <td>(</td>
          <td class="paramtype">const std::size_t</td>          <td class="paramname"><span class="paramname"><em>n</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Stat_ *const</td>          <td class="paramname"><span class="paramname"><em>statistic</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Bool_ *const</td>          <td class="paramname"><span class="paramname"><em>output</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Stat_</td><td>Type of the variance statistic. </td></tr>
    <tr><td class="paramname">Bool_</td><td>Type to be used as a boolean.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">statistic</td><td>Pointer to an array of length <code>n</code> containing the per-gene variance statistics. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array of length <code>n</code>. On output, this is filled with <code>true</code> if the gene is to be retained and <code>false</code> otherwise. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ad2e99f0413753c20e33c217e273ee120" name="ad2e99f0413753c20e33c217e273ee120"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad2e99f0413753c20e33c217e273ee120">&#9670;&#160;</a></span>choose_highly_variable_genes() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Bool_  = char, typename Stat_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; Bool_ &gt; scran_variances::choose_highly_variable_genes </td>
          <td>(</td>
          <td class="paramtype">const std::size_t</td>          <td class="paramname"><span class="paramname"><em>n</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Stat_ *const</td>          <td class="paramname"><span class="paramname"><em>statistic</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Bool_</td><td>Type to be used as a boolean. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Type of the variance statistic.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">statistic</td><td>Pointer to an array of length <code>n</code> containing the per-gene variance statistics. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A vector of booleans of length <code>n</code>, indicating whether each gene is to be retained. </dd></dl>

</div>
</div>
<a id="a526febe6eb07121cdd6649a938992ad2" name="a526febe6eb07121cdd6649a938992ad2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a526febe6eb07121cdd6649a938992ad2">&#9670;&#160;</a></span>choose_highly_variable_genes_index()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index_ , typename Stat_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; Index_ &gt; scran_variances::choose_highly_variable_genes_index </td>
          <td>(</td>
          <td class="paramtype">const Index_</td>          <td class="paramname"><span class="paramname"><em>n</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Stat_ *const</td>          <td class="paramname"><span class="paramname"><em>statistic</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ChooseHighlyVariableGenesOptions.html">ChooseHighlyVariableGenesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index_</td><td>Type of the indices. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Type of the variance statistic.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">statistic</td><td>Pointer to an array of length <code>n</code> containing the per-gene variance statistics. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Vector of sorted and unique indices for the chosen genes. All indices are guaranteed to be non-negative and less than <code>n</code>. </dd></dl>

</div>
</div>
<a id="a774f5059db93cedb24f461201f3131fd" name="a774f5059db93cedb24f461201f3131fd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a774f5059db93cedb24f461201f3131fd">&#9670;&#160;</a></span>fit_variance_trend() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Float_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void scran_variances::fit_variance_trend </td>
          <td>(</td>
          <td class="paramtype">const std::size_t</td>          <td class="paramname"><span class="paramname"><em>n</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float_ *const</td>          <td class="paramname"><span class="paramname"><em>mean</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float_ *const</td>          <td class="paramname"><span class="paramname"><em>variance</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Float_ *const</td>          <td class="paramname"><span class="paramname"><em>fitted</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Float_ *const</td>          <td class="paramname"><span class="paramname"><em>residuals</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structscran__variances_1_1FitVarianceTrendWorkspace.html">FitVarianceTrendWorkspace</a>&lt; Float_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>workspace</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1FitVarianceTrendOptions.html">FitVarianceTrendOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>We fit a trend to the per-feature variances against the means, both of which are computed from log-normalized expression data. We use a LOWESS smoother in several steps:</p>
<ol type="1">
<li>Filter out low-abundance genes, to ensure the span of the smoother is not skewed by many low-abundance genes.</li>
<li>Take the quarter-root of the variances, to squeeze the trend towards 1. This makes the trend more "linear" to improve the performance of the LOWESS smoother; it also reduces the chance of obtaining negative fitted values.</li>
<li>Apply the LOWESS smoother to the quarter-root variances. This is done using the implementation in the <b><a class="elRef" href="https://ltla.github.io/CppWeightedLowess/namespaceWeightedLowess.html">WeightedLowess</a></b> library.</li>
<li>Reverse the quarter-root transformation to obtain the fitted values for all non-low-abundance genes.</li>
<li>Extrapolate linearly from the left-most fitted value to the origin to obtain fitted values for the previously filtered genes. This is empirically justified by the observation that mean-variance trends of log-expression data are linear at very low abundances.</li>
</ol>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Float_</td><td>Floating-point type for the statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of features. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mean</td><td>Pointer to an array of length <code>n</code>, containing the means for all features. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">variance</td><td>Pointer to an array of length <code>n</code>, containing the variances for all features. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Pointer to an array of length <code>n</code>, to store the fitted values. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Pointer to an array of length <code>n</code>, to store the residuals. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">workspace</td><td>Collection of temporary data structures. This can be re-used across multiple <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code> calls. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ada30bbfa19a0c28c23fb69f61355ea54" name="ada30bbfa19a0c28c23fb69f61355ea54"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ada30bbfa19a0c28c23fb69f61355ea54">&#9670;&#160;</a></span>fit_variance_trend() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Float_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran__variances_1_1FitVarianceTrendResults.html">FitVarianceTrendResults</a>&lt; Float_ &gt; scran_variances::fit_variance_trend </td>
          <td>(</td>
          <td class="paramtype">const std::size_t</td>          <td class="paramname"><span class="paramname"><em>n</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float_ *const</td>          <td class="paramname"><span class="paramname"><em>mean</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float_ *const</td>          <td class="paramname"><span class="paramname"><em>variance</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1FitVarianceTrendOptions.html">FitVarianceTrendOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Overload of <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code> that allocates the output vectors.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Float_</td><td>Floating-point type for the statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of features. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mean</td><td>Pointer to an array of length <code>n</code>, containing the means for all features. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">variance</td><td>Pointer to an array of length <code>n</code>, containing the variances for all features. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Result of the trend fit, containing the fitted values and residuals for each gene. </dd></dl>

</div>
</div>
<a id="a34ffafc531b113e60ae57b3b5f8256be" name="a34ffafc531b113e60ae57b3b5f8256be"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a34ffafc531b113e60ae57b3b5f8256be">&#9670;&#160;</a></span>model_gene_variances_blocked() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void scran_variances::model_gene_variances_blocked </td>
          <td>(</td>
          <td class="paramtype">const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>mat</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *const</td>          <td class="paramname"><span class="paramname"><em>block</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedBuffers.html">ModelGeneVariancesBlockedBuffers</a>&lt; Stat_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>buffers</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking. The mean and variance of each gene is computed separately for all cells in each block, and a separate trend is fitted to each block to obtain residuals (see <code><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances()</a></code>). This ensures that sample and batch effects do not confound the variance estimates.</p>
<p>We also compute the average of each statistic across blocks, using the weighting strategy specified in <code><a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html#adcc6c3a8d53c6c7b785dedc53f7ef962">ModelGeneVariancesOptions::block_weight_policy</a></code>. The average residual is particularly useful for feature selection with <code><a class="el" href="#a836ab33e2b96134e9b3ad47df42aeb5e">choose_highly_variable_genes()</a></code>.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type to hold the block IDs. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>A <b>tatami</b> matrix containing log-expression values. Rows should be genes while columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells. Each entry should be a 0-based block identifier in \([0, B)\) where \(B\) is the total number of blocks. <code>block</code> can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">buffers</td><td>Collection of pointers of arrays in which to store the output statistics. The length of <code><a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedResults.html#a924b8308d55e630808c783b543ca89c4">ModelGeneVariancesBlockedResults::per_block</a></code> should be equal to the number of blocks. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="af48891a8979a6e46b37aaf7c9bc027cb" name="af48891a8979a6e46b37aaf7c9bc027cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af48891a8979a6e46b37aaf7c9bc027cb">&#9670;&#160;</a></span>model_gene_variances() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void scran_variances::model_gene_variances </td>
          <td>(</td>
          <td class="paramtype">const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>mat</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBuffers.html">ModelGeneVariancesBuffers</a>&lt; Stat_ &gt;</td>          <td class="paramname"><span class="paramname"><em>buffers</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Here, we scan through a log-transformed normalized expression matrix and compute per-gene means and variances. We then fit a trend to the variances with respect to the means using <code><a class="el" href="#a774f5059db93cedb24f461201f3131fd">fit_variance_trend()</a></code>. We assume that most genes at any given abundance are not highly variable, such that the fitted value of the trend is interpreted as the "uninteresting" variance - this is mostly attributed to technical variation like sequencing noise, but can also represent constitutive biological noise like transcriptional bursting. Under this assumption, the residual can be treated as a measure of biologically interesting variation, and can be used to identify relevant features for downstream analyses.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>A <b>tatami</b> matrix containing log-expression values. Rows should be genes while columns should be cells. </td></tr>
    <tr><td class="paramname">buffers</td><td>Collection of buffers in which to store the computed statistics. </td></tr>
    <tr><td class="paramname">options</td><td>Further options. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ab7571d9531d24fb47fa0874ff5039725" name="ab7571d9531d24fb47fa0874ff5039725"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab7571d9531d24fb47fa0874ff5039725">&#9670;&#160;</a></span>model_gene_variances() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Stat_  = double, typename Value_ , typename Index_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran__variances_1_1ModelGeneVariancesResults.html">ModelGeneVariancesResults</a>&lt; Stat_ &gt; scran_variances::model_gene_variances </td>
          <td>(</td>
          <td class="paramtype">const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>mat</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Overload of <code><a class="el" href="#af48891a8979a6e46b37aaf7c9bc027cb">model_gene_variances()</a></code> that allocates space for the output statistics.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics. </td></tr>
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>A <b>tatami</b> matrix containing log-expression values. Rows should be genes while columns should be cells. </td></tr>
    <tr><td class="paramname">options</td><td>Further options.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Results of the variance modelling. </dd></dl>

</div>
</div>
<a id="af13889165f096e5af5db48a6046e2d27" name="af13889165f096e5af5db48a6046e2d27"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af13889165f096e5af5db48a6046e2d27">&#9670;&#160;</a></span>model_gene_variances_blocked() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Stat_  = double, typename Value_ , typename Index_ , typename Block_ &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran__variances_1_1ModelGeneVariancesBlockedResults.html">ModelGeneVariancesBlockedResults</a>&lt; Stat_ &gt; scran_variances::model_gene_variances_blocked </td>
          <td>(</td>
          <td class="paramtype">const <a class="elRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a>&lt; Value_, Index_ &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>mat</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *const</td>          <td class="paramname"><span class="paramname"><em>block</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html">ModelGeneVariancesOptions</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>options</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Overload of <code><a class="el" href="#a34ffafc531b113e60ae57b3b5f8256be">model_gene_variances_blocked()</a></code> that allocates space for the output statistics.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics. </td></tr>
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>A <b>tatami</b> matrix containing log-expression values. Rows should be genes while columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing 0-based block identifiers. This may also be a <code>nullptr</code> in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">options</td><td>Further options.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Results of the variance modelling in each block. An average for each statistic is also computed if <code><a class="el" href="structscran__variances_1_1ModelGeneVariancesOptions.html#a7817148b82c400598b98804e9beb6d56">ModelGeneVariancesOptions::compute_average</a> = true</code>. </dd></dl>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.12.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
